File: backend/app/core/wine_types.py
import re

CANONICAL_WINE_TYPES = ("Red", "White", "Rose", "Sparkling", "Sweet", "Fortified", "Other")


def normalize_wine_type(value: str | None) -> str:
    cleaned = (value or "").strip()
    if not cleaned:
        return ""
    normalized = cleaned.lower()
    if normalized in {"red", "rosso"} or "vino rosso" in normalized:
        return "Red"
    if normalized in {"white", "bianco"} or "vino bianco" in normalized:
        return "White"
    if normalized in {"rose", "ros\u00e9", "rosato"}:
        return "Rose"
    if normalized in {"sparkling", "spumante", "champagne"}:
        return "Sparkling"
    if normalized in {"sweet", "dolce"}:
        return "Sweet"
    if normalized in {"fortified", "fortificato"}:
        return "Fortified"
    if normalized in {"other", "altro"}:
        return "Other"
    return next((item for item in CANONICAL_WINE_TYPES if item.lower() == normalized), cleaned)


def normalize_verified_vintage(value: str | None) -> str:
    """Accept only an explicit vintage year or the standard non-vintage markers."""
    cleaned = (value or "").strip().upper()
    return cleaned if re.fullmatch(r"(?:19|20)\d{2}|NV|MV", cleaned) else ""

File: frontend/src/domain/wineTypes.ts
export const canonicalWineTypes = ["Red", "White", "Rose", "Sparkling", "Sweet", "Fortified", "Other"] as const;

export function normalizeWineType(value: string | null | undefined) {
  const trimmed = (value || "").trim();
  if (!trimmed) return "";

  const normalized = trimmed
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "");

  if (["red", "rosso"].includes(normalized) || normalized.includes("vino rosso")) return "Red";
  if (["white", "bianco"].includes(normalized) || normalized.includes("vino bianco")) return "White";
  if (["rose", "rosato"].includes(normalized)) return "Rose";
  if (["sparkling", "spumante", "champagne"].includes(normalized)) return "Sparkling";
  if (["sweet", "dolce"].includes(normalized)) return "Sweet";
  if (["fortified", "fortificato"].includes(normalized)) return "Fortified";
  if (["other", "altro"].includes(normalized)) return "Other";

  const canonical = canonicalWineTypes.find((type) => type.toLowerCase() === normalized);
  return canonical || trimmed;
}
